verification.md

```diff
@@ -11,24 +12,296 @@ module VERIFICATION
+              : nthbyteof(V,  2, 32)
+              : nthbyteof(V,  3, 32)
+              : nthbyteof(V,  4, 32)
+              : nthbyteof(V,  5, 32)
+              : nthbyteof(V,  6, 32)
+              : nthbyteof(V,  7, 32)
+              : nthbyteof(V,  8, 32)
+              : nthbyteof(V,  9, 32)
+              : nthbyteof(V, 10, 32)
+              : nthbyteof(V, 11, 32)
+              : nthbyteof(V, 12, 32)
+              : nthbyteof(V, 13, 32)
+              : nthbyteof(V, 14, 32)
+              : nthbyteof(V, 15, 32)
+              : nthbyteof(V, 16, 32)
+              : nthbyteof(V, 17, 32)
+              : nthbyteof(V, 18, 32)
+              : nthbyteof(V, 19, 32)
+              : nthbyteof(V, 20, 32)
+              : nthbyteof(V, 21, 32)
+              : nthbyteof(V, 22, 32)
+              : nthbyteof(V, 23, 32)
+              : nthbyteof(V, 24, 32)
+              : nthbyteof(V, 25, 32)
+              : nthbyteof(V, 26, 32)
+              : nthbyteof(V, 27, 32)
+              : nthbyteof(V, 28, 32)
+              : nthbyteof(V, 29, 32)
+              : nthbyteof(V, 30, 32)
+              : nthbyteof(V, 31, 32)
+              : .WordStack ) => V
+    requires 0 <=Int V andBool V <Int (2 ^Int 256)
+
+  rule #padToWidth(32, #asByteStack(V)) => #asByteStackInWidth(V, 32)
+    requires 0 <=Int V andBool V <Int (2 ^Int 256)
+
+  // syntactic sugar
+  syntax WordStack ::= #asByteStackInWidth(Int, Int) [function]
+                     | #asByteStackInWidthaux(Int, Int, Int, WordStack) [function]
+  rule #asByteStackInWidth(X, N) => #asByteStackInWidthaux(X, N -Int 1, N, .WordStack)
+  rule #asByteStackInWidthaux(X, I, N, WS) => #asByteStackInWidthaux(X, I -Int 1, N, nthbyteof(X, I, N) : WS) when I >Int 0
+  rule #asByteStackInWidthaux(X, 0, N, WS) => nthbyteof(X, 0, N) : WS
+```
+
+Another type of byte-array manipulating operation is used to extract the function signature from the call data. The function signature is located in the first four bytes of the call data, but there is no atomic EVM instruction that can load only the four bytes, thus some kind of byte-twiddling operations are necessary.
+
+The extraction mechanism varies by language compilers. For example, in Viper, the first 32 bytes of the call data are loaded into the memory at the starting location 28 (i.e., in the memory range of 28 to 59), and the memory range of 0 to 31, which consists of 28 zero bytes and the four signature bytes, is loaded into the stack. In Solidity, however, the first 32 bytes of the call data are loaded into the stack, and the loaded word (i.e., a 256-bit integer) is divided by `2^(28*8)` (i.e., left-shifted by 28 bytes), followed by masked by 0xffffffff (i.e., 4 bytes of bit 1’s).
+
+The following lemmas essentially capture the signature extraction mechanisms. It reduces the reasoning efforts of the underlying theorem prover, factoring out the essence of the byte-twiddling operations.
+
+```{.k .uiuck}
+  // for viper
```

Comment:
should we put the rule for Viper vs. for Solidity in different  modules/files?

Reply:
I thought the similar thing. We can discuss more about it with others.

I wish I can put this rules in the spec file, but cannot due to the limitation of k prover's parsing mechanism.

Reply:
report this issue to @dwightguth .  He may have a quick fix for us.

Reply:
I'm not positive about it. It's quite complicated and messy. But will report it.

Reply:
@grosu Not entire sure if indeed it is good to do. If we want to extend the semantics, (even if it is just about defining DSLs that are nothing to do with the semantics itself), we still want to test it to make sure it doesn't screw up. In that sense, it is better to have separate modules for each language/DSL, but then, we have to kompile multiple version of extended semantics and test them every time, which may be too much overhead.